Fix for link color in containers

oandregal opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/33437

When we [landed support for elements](https://github.com/WordPress/gutenberg/pull/31524#discussion_r634252402) (the link color) we made it so the generated class was this:

`.wp-element-id a { color: color-value !important; }`

Notice the `!important`. This was to make sure user-provided styles weren't overridden by theme-provided styles. However, the selector is too far-reaching and it's affecting blocks that use the `a` but, arguably, should not be considered "links" (such as button and social links).

This PR removes the `!important` so the styles for the link color are:

`.wp-element-id a { color: color-value; }`

**If we go ahead with this change, themes that add additional CSS for link colors need to make sure that user-provided styles are respected in every situation.**

## Testing

<details>
<summary>Blocks at the top-level</summary>

I tested this with the TT1-blocks:

- A paragraph with a link without any user color provided.
- A paragraph with a link and user-provided text color. Make sure links are unaffected by the user-provided text color.
- A paragraph with a link and user-provided text & link colors. Make sure links are affected by the user-provided link color.

![tt1-blocks-top-level](https://user-images.githubusercontent.com/583546/132674665-4c540605-d2a1-4c57-8fa3-7f4ae1d623eb.png)

Note that the **TwentyTwentyOne theme needs to be updated** to not override user styles (see the last paragraph, which should show the links in red):

![twentytwentyone-top-level](https://user-images.githubusercontent.com/583546/132674748-9b26ce59-0afd-4860-9110-b104261a1f7e.png)
</details>

<details>
<summary>Blocks within containers</summary>

I tested this with the TwentyTwentyOne theme:

- Add a group that contains a paragraph with a link, the button block, and the social links. Do not assign any color to the group.
- Add a group that contains a paragraph with a link, the button block, and the social links. Set the text color for the group. Make sure links are unaffected by the text color.
- Add a group that contains a paragraph with a link, the button block, and the social links. Set the text and link color for the group. Make sure links are affected by the link color.

![twentytwentyone-group](https://user-images.githubusercontent.com/583546/132675137-8055363c-649d-42a2-a579-d3b6ab0cfa55.png)

Note that the **TT1-blocks theme needs to be updated** to make sure link color is not used in the button block:

![tt1-blocks-group](https://user-images.githubusercontent.com/583546/132675241-a5852e55-0881-4651-a221-eb6a4d21d4b8.png)
</details>

<details>
<summary>Nested blocks</summary>

See https://github.com/WordPress/gutenberg/pull/34689#issuecomment-916167610 for testing instructions.

</details>

oandregal committed:
Remove important rule

oandregal committed:
Contain specificity of links rule so themes can override it

oandregal commented on lib/block-supports/elements.php:
```diff
@@ -38,7 +38,7 @@ function gutenberg_render_elements_support( $block_content, $block ) {
 	}
 	$link_color_declaration = esc_html( safecss_filter_attr( "color: $link_color" ) );
 
-	$style = "<style>.$class_name a{" . $link_color_declaration . " !important;}</style>\n";
```
This is going to need changes in the tests. Will update them once we're certain this is the right direction.

oandregal replied:
This is going to have 011 specificity.

I've looked at all current cases of core blocks and this change will still work well with global styles. The reason is that the current global styles generated for links have 011 or 002, depending on whether they are `.wp-block-* a` or `element a`. The element styles are loaded after the global styles, so the element styles win.

I was thinking that we could add the `:not(.has-link-color)` selector to the global styles we generate. Should any block have higher than 011 specificity this selector will prevent it from applying. However! This is also not complete: it won't work for some blocks that declare `__experimentalSelector` and target an inner element. Use case:

- some block has a wrapper such as `.wp-block-wrapper`
- for whatever reason, the block declares that the styles should match some inner element instead `__experimentalSelector: .wp-block-wrapper-inner .wp-block-wrapper-extra`

Adding the negation in this situation would render `.wp-block-wrapper-inner .wp-block-wrapper-extra:not(.has-link-color) { /* link color */ }` which has higher specificity that the link color element and still targets the link in the block because the `.has-link-color` class lives in the wrapper.

getdave reviewed (commented):
Thank you for working on this 🏅 

I don't feel in a position to approve this as I don't know enough about the implications.

However, the original case which I raised is now fixed.

<img width="1174" alt="Screen Shot 2021-09-09 at 12 39 15" src="https://user-images.githubusercontent.com/444434/132679261-57ec1ec4-87bf-43d9-8bba-e886a2f35eed.png">
<img width="1403" alt="Screen Shot 2021-09-09 at 12 38 48" src="https://user-images.githubusercontent.com/444434/132679272-ede6946b-c20f-4859-a6f4-92be7d36bb86.png">

ryelle commented:
This fixes the issue in the patterns I flagged, so thanks!

> Themes that add additional CSS for link colors need to make sure that user-provided styles are respected in every situation.

Link color is still an experimental feature, so changing the behavior is to be expected, right? Or is this something that will need to be documented?

I did manage to find one more edge case, where a nested link color is overridden by a link color on the parent container, but only on the frontend.

| Editor | Front end |
|-------|----------|
| <img width="673" alt="Screen Shot 2021-09-09 at 10 19 44 AM" src="https://user-images.githubusercontent.com/541093/132705414-790ac5f4-11a8-4b36-b5de-ef0379276a26.png"> | <img width="797" alt="Screen Shot 2021-09-09 at 10 19 35 AM" src="https://user-images.githubusercontent.com/541093/132705434-df7da95a-55f9-4e4b-b539-5288c5db4920.png"> |

The group block has a white link color, but that individual paragraph uses a bright orange. In the editor, the `<style>…` is output before the block it applies to, but on the front end, the styles are output after the block. So the white link rule comes "after" the orange rule, and overrides it.

In the editor…
```css
/* Before the group `div` */
.wp-elements-613a1a9b8f7d4 a {
    color: var(--wp--preset--color--white);
}
/* Before the `p` */
.wp-elements-613a1a9b8f45f a {
    color: #cf6100;
}
```

Link is orange ✅ 

On the front end…
```css
/* After the `p` */
.wp-elements-613a1a9b8f45f a {
    color: #cf6100;
}
/* After the group `div` */
.wp-elements-613a1a9b8f7d4 a {
    color: var(--wp--preset--color--white);
}
```

Link is white ❌

oandregal committed:
Remove important from the inline styles as well

oandregal committed:
Output styles before the element

The reason for this is nested content such as:

    <group class="wp-element-1">
      <p>Text with <link> and colors inherited from group</p>
      <p class="wp-element-2">Text with <link> and own colors</p>
    </group>

The selectors for the link are

- ".wp-element 1 a" => generated by the group
- ".wp-element-2 a" => generated by the paragraph

because they have the same specificity the latest will win.
We need the styles to respect the DOM order of the elements
that have them.

    <style .wp-element-1>
    <style .wp-element-2>
    <group ...>...</group>

oandregal commented:
Nice one @ryelle ! Pushed a fix at https://github.com/WordPress/gutenberg/pull/34689/commits/8d38f00b3aaf6f75fe1321c35e4f54eef58ac8c5

oandregal commented:
> Link color is still an experimental feature, so changing the behavior is to be expected, right? Or is this something that will need to be documented?

It needs to be documented. For themes that use the `theme.json` is not experimental.

ryelle commented:
Now it works in all the scenarios I could think of 👍🏻

Mamaduka reviewed (commented):
@oandregal, I think the snapshot in PHP unit tests needs an update.

oandregal commented:
ok, this PR may work at a practical level. Theoretically, this could break in some situations https://github.com/WordPress/gutenberg/pull/34689#discussion_r705233508

What I don't like about this is that themes will need to fix some CSS and that they need to take care of not overriding user-provided styles. However, the current situation is also not ideal as blocks that use the `a` element are affected (button and cover, for example).

I'm going to let this sit for a few days in case we can come with an alternative that fixes the issues and doesn't have the drawbacks. If we don't, I'll update the tests and merge this.

cc @jorgefilipecosta @jasmussen in case you have thoughts.

jasmussen commented:
> What I don't like about this is that themes will need to fix some CSS and that they need to take care of not overriding user-provided styles. However, the current situation is also not ideal as blocks that use the a element are affected (button and cover, for example).

Themes will always be able to break things, whether for good or for bad. In this case, the removal of !important's feels worth any followup we might have to do. Thanks for the PR.

jorgefilipecosta reviewed (approved):
It seems to work well and I think in the long term it is better to not use important. We should just create a dev not and let themes beware of this change in case link color stops working as expected on some theme.

getsource commented:
Thanks so much for your work on this @oandregal !
I agree with the comments re: leaning towards removing the `!important` if it's possible.

I'm guessing this applies to some of the comments about specific theme issues:
I noticed that when testing this PR with Twenty Twenty-One, in the editor, paragraph and group link color overrides weren't reflected, but they displayed properly when rendered on the front-end.

oandregal commented:
Thanks all for the help. I'm going to fix the tests for this one and will merge it.

oandregal committed:
Update link element tests

getdave commented:
@oandregal FYI I tried to build a block theme over the weekend and ran into this `!important` issue several times. I'm very glad it will be removed.

oandregal commented:
Rebased to get https://github.com/WordPress/gutenberg/pull/34776 in (a potential fix for a failing e2e test)

oandregal commented:
## Devnote

**_To review/clarify/check when time comes to publish it:_**

We've made some changes to the link color feature that affect its specificity and helped to fix some issues with links within container blocks:

- We've lowered the specificity of the link color rule in the block-library package. The selector is now `:where(p.has-text-color:not(.has-link-color)) a`, which resolves to 001 specificity.
- By doing so, we could lower the global styles specifitiy as well by no longer making it  `!important` (only affects themes with `theme.json`). The global link uses the `a` selector and the link property within blocks uses `<block-selector> a`. By loading after the block-styles stylesheet, they'll have higher priority.
- When users select a color for the link, two things happen: the block wrapper gets assigned the class `has-link-color` and a new `.wp-element-ID` class. The styles for the link color are generated dynamically using the `.wp-element-ID a` selector, which is 011 specificity.

Given these changes, the theme needs to make sure its styles play well with user choices.
